### PR TITLE
JU-116 - Java 21 Migration

### PR DESCRIPTION
**Pull request by aduffy-ch**
https://companieshouse.atlassian.net/browse/JU-116

Make code changes and library changes to Java v:21

Clearing critical vulnerabilities (if other vulnerabilities can be resolved, then do so)

Code cleanup (clear configs specific to Mesos & update docker images)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -9,37 +9,39 @@
...
+		<java.version>21</java.version>
 
 		<main.class>uk.gov.companieshouse.charges.data.ChargesDataApiApplication</main.class>
 
 		<maven.compiler.source>${java.version}</maven.compiler.source>
 		<maven.compiler.target>${java.version}</maven.compiler.target>
 		<maven-compiler-plugin.version>3.8.1</maven-compiler-plugin.version>
-		<spring.boot.version>2.6.4</spring.boot.version>
+		<spring.boot.version>3.2.3</spring.boot.version>
 		<assertj-core.version>3.22.0</assertj-core.version>
```
Comment (nhill-ch) [inline pom.xml:29]: can this be brought up to 3.25.3

File: pom.xml [hunk 2]
```diff
@@ -9,37 +9,39 @@
 	</parent>
 
 	<artifactId>charges-data-api</artifactId>
+
+
```
Comment (nhill-ch) [inline pom.xml:13]: remove extra whitespace

File: pom.xml [hunk 3]
```diff
@@ -9,37 +9,39 @@
...
 		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
 		<project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
-		<java.version>11</java.version>
+		<java.version>21</java.version>
 
 		<main.class>uk.gov.companieshouse.charges.data.ChargesDataApiApplication</main.class>
 
 		<maven.compiler.source>${java.version}</maven.compiler.source>
 		<maven.compiler.target>${java.version}</maven.compiler.target>
 		<maven-compiler-plugin.version>3.8.1</maven-compiler-plugin.version>
```
Comment (nhill-ch) [inline pom.xml:27]: 3.12.1?

File: pom.xml [hunk 4]
```diff
@@ -9,37 +9,39 @@
...
+		<structured-logging.version>3.0.3</structured-logging.version>
+		<api-sdk-manager-java-library.version>3.0.5</api-sdk-manager-java-library.version>
+		<api-helper-java-library.version>3.0.1</api-helper-java-library.version>
+		<cucumber-java.version>7.15.0</cucumber-java.version>
+		<jib-maven-plugin.version>3.4.0</jib-maven-plugin.version>
 		<skip.integration.tests>false</skip.integration.tests>
 		<skip.unit.tests>false</skip.unit.tests>
-		<private-api-sdk-java.version>2.0.181</private-api-sdk-java.version>
+		<private-api-sdk-java.version>4.0.76</private-api-sdk-java.version>
 		<io-cucumber.version>7.2.3</io-cucumber.version>
```
Comment (nhill-ch) [inline pom.xml:43]: 7.15.0

File: pom.xml [hunk 5]
```diff
@@ -9,37 +9,39 @@
...
-		<wiremock.standalone.version>2.32.0</wiremock.standalone.version>
+		<wiremock.standalone.version>3.0.1</wiremock.standalone.version>
 		<maven-resources-plugin.version>3.2.0</maven-resources-plugin.version>
 
 		<!--sonar configuration-->
```
Comment (nhill-ch) [inline pom.xml:47]: 3.11.0.3922

File: pom.xml [hunk 6]
```diff
@@ -210,16 +206,27 @@
...
 		<dependency>
-			<groupId>org.springframework.boot</groupId>
-			<artifactId>spring-boot-starter-security</artifactId>
+			<groupId>org.springframework</groupId>
+			<artifactId>spring-test</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>io.cucumber</groupId>
+			<artifactId>cucumber-java</artifactId>
+			<version>7.8.1</version>
```
Comment (nhill-ch) [inline pom.xml:226]: use ${io-cucumber.version}

**Review by nhill-ch — CHANGES_REQUESTED** [review]
Added a few comments about upgrading some versions in the pom file, haven't added all, please go back and upgrade any which haven't been upgraded

File: pom.xml (hunk 6 above)
Comment (aduffy-ch) [reply pom.xml:226]: done

File: pom.xml [hunk 7]
```diff
@@ -308,7 +315,7 @@
 			<plugin>
 				<groupId>org.codehaus.mojo</groupId>
 				<artifactId>build-helper-maven-plugin</artifactId>
-				<version>${maven-build-helper-plugin.version}</version>
+				<version>3.5.0</version>
```
[the remark this answers is not in the document]
Comment (aduffy-ch) [reply pom.xml:318]: done

File: pom.xml (hunk 2 above)
Comment (aduffy-ch) [reply pom.xml:13]: done

File: pom.xml (hunk 4 above)
Comment (aduffy-ch) [reply pom.xml:43]: done

File: src/main/java/uk/gov/companieshouse/charges/data/config/WebSecurityConfig.java [hunk 8]
```diff
@@ -1,50 +1,46 @@
...
     /**
      * Configure Web Security.
      */
-    @Override
-    public void configure(WebSecurity web) throws  Exception {
+    @Bean
+    public WebSecurityCustomizer webSecurityCustomizer() {
         // Excluding healthcheck endpoint from security filter
-        web.ignoring().antMatchers("/charges-data-api/healthcheck");
+        return web -> web.ignoring().requestMatchers("/healthcheck");
```
[the remark this answers is not in the document]
Comment (aduffy-ch) [reply src/main/java/uk/gov/companieshouse/charges/data/config/WebSecurityConfig.java:44]: done

File: src/main/java/uk/gov/companieshouse/charges/data/auth/EricTokenAuthenticationFilter.java [hunk 9]
```diff
@@ -19,7 +19,6 @@ public EricTokenAuthenticationFilter(Logger logger) {
         this.logger = logger;
     }
 
-    @Override
```
[the remark this answers is not in the document]
Comment (aduffy-ch) [reply src/main/java/uk/gov/companieshouse/charges/data/auth/EricTokenAuthenticationFilter.java:22]: done

File: pom.xml (hunk 1 above)
Comment (aduffy-ch) [reply pom.xml:29]: done

File: pom.xml (hunk 3 above)
Comment (aduffy-ch) [reply pom.xml:27]: done

File: pom.xml (hunk 5 above)
Comment (aduffy-ch) [reply pom.xml:47]: is this comment for the wiremock.standalone.version?

**Review by hwatkins-ch — APPROVED** [review]

File: pom.xml (hunk 5 above)
Comment (nhill-ch) [reply pom.xml:45]: no the sonar-maven-plugin

**Review by nhill-ch — CHANGES_REQUESTED** [review]
things that still need to get upgraded:
jackson-datatype-jsr310
sonar-maven-plugin
maven-checkstyle-plugin
maven-resources-plugin
com.puppycrawl.tools

File: pom.xml (hunk 5 above)
Comment (aduffy-ch) [reply pom.xml:45]: done

**Review by nhill-ch — APPROVED** [review]

**Pull request merged**